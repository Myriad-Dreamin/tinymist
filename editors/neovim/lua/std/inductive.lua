---@generic S
---@alias Constructor fun(self: any, ...: any): S
---@alias InductiveMethod fun(self: S, ...: any): any

---@alias ConstructorDefs table<string, Constructor> | table<string, table<string, InductiveMethod>>

---@class Inductive<S> : { [string]: InductiveMethod }
---@operator call(table): `S`

---Create a new inductive type.
---@param name string The name of the new type, used only for errors
---@param defs ConstructorDefs A table of constructor definitions
---@return Inductive
return function(name, defs)
  local Type = {}

  local to_obj

  local _, first = next(defs)
  if type(first) ~= 'table' then
    to_obj = function(_, t)
      return t
    end
  else
    local methods = vim.tbl_keys(first)

    to_obj = function(constructor_name, impl)
      local obj = setmetatable({
        serialize = function(self)
          return { [constructor_name] = self[1] }
        end,
      }, { __index = Type })

      for _, method_name in ipairs(methods) do
        local method = impl[method_name]

        if not method then
          error(('%s method is missing for %s.%s'):format(method_name, name, constructor_name))
        end
        obj[method_name] = method
        impl[method_name] = nil -- so we can tell if there are any extras...
      end

      local extra = next(impl)
      if extra then
        error(('%s method is unexpected for %s.%s'):format(extra, name, constructor_name))
      end
      return function(_, ...)
        return setmetatable({ ... }, {
          __index = obj,
          __call = function(_, ...)
            return Type[constructor_name](Type, ...)
          end,
        })
      end
    end
  end

  for constructor_name, impl in pairs(defs) do
    Type[constructor_name] = to_obj(constructor_name, impl)
  end
  return setmetatable(Type, {
    __call = function(self, data, ...)
      local constructor_name, value = next(data)
      local constructor = self[constructor_name]
      if not constructor then
        error(('Invalid %s constructor: %s'):format(name, constructor_name))
      end
      return constructor(self, value, ...)
    end,
  })
end
