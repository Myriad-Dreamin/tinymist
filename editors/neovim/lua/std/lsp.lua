local lsp = {}

---Convert an LSP position to a (0, 0)-indexed tuple.
---
---These are used by extmarks.
---See `:h api-indexing` for details.
---@param position lsp.Position
---@param bufnr integer the buffer whose position is referred to
---@return { [1]: integer, [2]: integer } position
function lsp.position_to_byte0(position, bufnr)
  local line = vim.api.nvim_buf_get_lines(bufnr, position.line, position.line + 1, false)[1] or ''
  local ok, col = pcall(vim.str_byteindex, line, position.character, true)
  return { position.line, ok and col or position.character }
end

---Polyfill vim.fs.relpath-ish for Neovim < 0.11.
---
---Don't use this for real filesystem operations (as opposed to display),
---its implementation is naive!
local relpath = vim.fs.relpath
  or function(base, target)
    if vim.startswith(target, base .. '/') then
      return target:sub(#base + 2)
    end
    return target
  end

---Convert LSP document params inside the current buffer to a human-readable (1, 1)-indexed string.
---
---Takes the workspace into account in order to return a relative path.
---@param params UIParams
function lsp.text_document_position_to_string(params)
  local workspace = vim.lsp.buf.list_workspace_folders()[1] or vim.uv.cwd()
  local filename = vim.uri_to_fname(params.textDocument.uri)

  return ('%s at %d:%d'):format(
    relpath(workspace, filename) or filename,
    params.position.line + 1,
    params.position.character + 1
  )
end

---Convert an LSP range to a human-readable (1, 1)-indexed string.
---
---We use 1-based indexing here as the `gg` and `|` motions are 1-indexed,
---which is the most likely way a human (you?) will interact with this
---information.
---@param range lsp.Range
function lsp.range_to_string(range)
  return ('%d:%d-%d:%d'):format(
    range.start.line + 1,
    range.start.character + 1,
    range['end'].line + 1,
    range['end'].character + 1
  )
end

-- ~*~ vim.lsp._private_functions we still need... ~*~

local format_line_ending = {
  ['unix'] = '\n',
  ['dos'] = '\r\n',
  ['mac'] = '\r',
}

---@private
---@param bufnr (number)
---@return string
local function buf_get_line_ending(bufnr)
  return format_line_ending[vim.bo[bufnr].fileformat] or '\n'
end

---@private
---Returns full text of buffer {bufnr} as a string.
---
---@param bufnr (number) Buffer handle, or 0 for current.
---@return string # Buffer text as string.
function lsp.buf_get_full_text(bufnr)
  local line_ending = buf_get_line_ending(bufnr)
  local text = table.concat(vim.api.nvim_buf_get_lines(bufnr, 0, -1, true), line_ending)
  if vim.bo[bufnr].eol then
    text = text .. line_ending
  end
  return text
end

-- vim.lsp.diagnostic has a *private* `diagnostic_lsp_to_vim` :/ ...
--
-- the below comes from there / is required for assembling vim.Diagnostic
-- objects out of LSP responses

---@param severity lsp.DiagnosticSeverity
function lsp.severity_lsp_to_vim(severity)
  if type(severity) == 'string' then
    severity = vim.lsp.protocol.DiagnosticSeverity[severity] ---@type integer
  end
  return severity
end

---@param diagnostic lsp.Diagnostic
---@param client_id integer
---@return table?
function lsp.tags_lsp_to_vim(diagnostic, client_id)
  local tags ---@type table?
  for _, tag in ipairs(diagnostic.tags or {}) do
    if tag == vim.lsp.protocol.DiagnosticTag.Unnecessary then
      tags = tags or {}
      tags.unnecessary = true
    elseif tag == vim.lsp.protocol.DiagnosticTag.Deprecated then
      tags = tags or {}
      tags.deprecated = true
    else
      vim.lsp.log.info(string.format('Unknown DiagnosticTag %d from LSP client %d', tag, client_id))
    end
  end
  return tags
end

return lsp
