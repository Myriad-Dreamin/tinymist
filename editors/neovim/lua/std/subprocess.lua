local subprocess = {}

---Run a subprocess, blocking on exit, and returning its stdout.
---@return string: the lines of stdout of the exited process
function subprocess.check_output(...)
  local process = vim.system(...)
  local result = process:wait()
  if result.code == 0 then
    return result.stdout
  end

  error(
    string.format(
      '%s exited with non-zero exit status %d.\nstderr contained:\n%s',
      vim.inspect(process.cmd),
      result.code,
      result.stderr
    )
  )
end

return subprocess
