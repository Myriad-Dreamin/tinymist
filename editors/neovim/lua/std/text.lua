local text = {}

local function max_common_indent(str)
  local _, _, common_indent, rest = str:find '^(%s*)(.*)'
  local common_indent_len = #common_indent
  local len
  for indent in rest:gmatch '\n( +)' do
    len = #indent
    if len < common_indent_len then
      common_indent, common_indent_len = indent, len
    end
  end
  return common_indent
end

---Dedent a multi-line string.
---@param str string
function text.dedent(str)
  str = str:gsub('\n *$', '\n') -- trim leading/trailing space
  local prefix = max_common_indent(str)
  return str:gsub('^' .. prefix, ''):gsub('\n' .. prefix, '\n')
end

---Build a single-line string out a multiline one, replacing \n with spaces.
function text.s(str)
  return text.dedent(str):gsub('\n', ' ')
end

return text
