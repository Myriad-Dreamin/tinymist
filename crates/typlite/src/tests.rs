mod model;
mod rendering;

use std::sync::{LazyLock, OnceLock};

use regex::Regex;
use tinymist_world::{CompileFontArgs, EntryState, FontResolverImpl, LspUniverseBuilder};
use typst_syntax::Source;

use super::*;

fn conv(s: &str) -> EcoString {
    static FONT_RESOLVER: LazyLock<Result<Arc<FontResolverImpl>>> = LazyLock::new(|| {
        Ok(Arc::new(
            LspUniverseBuilder::resolve_fonts(CompileFontArgs::default())
                .map_err(|e| format!("{e:?}"))?,
        ))
    });

    let font_resolver = FONT_RESOLVER.clone();
    let cwd = std::env::current_dir().unwrap();
    let main = Source::detached(s);
    let mut universe = LspUniverseBuilder::build(
        EntryState::new_rooted(cwd.as_path().into(), Some(main.id())),
        font_resolver.unwrap(),
        Default::default(),
        Default::default()
    )
    .unwrap();
    universe
        .map_shadow_by_id(main.id(), Bytes::from(main.text().as_bytes().to_owned()))
        .unwrap();
    let world = universe.snapshot();

    let res = Typlite::new(Arc::new(world)).convert().unwrap();
    static REG: OnceLock<Regex> = OnceLock::new();
    let reg = REG.get_or_init(|| Regex::new(r#"data:image/svg\+xml;base64,([^"]+)"#).unwrap());
    let res = reg.replace(&res, |_captures: &regex::Captures| {
        // let hash = _captures.get(1).unwrap().as_str();
        // format!(
        //     "data:image-hash/svg+xml;base64,siphash128:{:x}",
        //     typst_shim::utils::hash128(hash)
        // )
        "data:image-hash/svg+xml;base64,redacted"
    });

    res.into()
}

#[test]
fn test_converted() {
    insta::assert_snapshot!(conv(r###"
= Hello, World!
This is a typst document.
        "###), @r###"
        # Hello, World!
        This is a typst document.
        "###);
    insta::assert_snapshot!(conv(r###"
Some inlined raw `a`, ```c b```
        "###), @"Some inlined raw `a`, `b`");
    insta::assert_snapshot!(conv(r###"
- Some *item*
- Another _item_
        "###), @r###"
        - Some **item**
        - Another _item_
        "###);
    insta::assert_snapshot!(conv(r###"
+ A
+ B
        "###), @r###"
        1. A
        1. B
        "###);
    insta::assert_snapshot!(conv(r###"
2. A
+ B
        "###), @r###"
        2. A
        1. B
        "###);
    insta::assert_snapshot!(conv(r###"
$
1/2 + 1/3 = 5/6
$
        "###), @r###"<p align="center"><img src="data:image-hash/svg+xml;base64,redacted" alt="typst-block" /></p>"###);
}
