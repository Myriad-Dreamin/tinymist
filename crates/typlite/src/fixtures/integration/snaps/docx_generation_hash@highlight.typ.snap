---
source: crates/typlite/src/tests.rs
expression: hash
input_file: crates/typlite/src/fixtures/integration/highlight.typ
snapshot_kind: text
---
siphash128_13:6e6d759532b382b4f775e85f9599d60d
